## Question 1:
Install Spark and PySpark

Install Spark
Run PySpark
Create a local spark session
Execute spark.version.
What's the output?

**3.3.2**

In [2]:
import pyspark
from pyspark.sql import SparkSession
from pyspark.sql import types
import pandas as pd
from pyspark.sql import functions as F

In [3]:
spark = SparkSession.builder \
    .master("local[*]") \
    .appName('test') \
    .getOrCreate()

Setting default log level to "WARN".
To adjust logging level use sc.setLogLevel(newLevel). For SparkR, use setLogLevel(newLevel).


24/02/24 11:25:49 WARN NativeCodeLoader: Unable to load native-hadoop library for your platform... using builtin-java classes where applicable
24/02/24 11:25:50 WARN Utils: Service 'SparkUI' could not bind on port 4040. Attempting port 4041.


In [3]:
spark.version

'3.3.2'

## Question 2:

FHV October 2019

Read the October 2019 FHV into a Spark Dataframe with a schema as we did in the lessons.
Repartition the Dataframe to 6 partitions and save it to parquet.
What is the average size of the Parquet (ending with .parquet extension) Files that were created (in MB)? Select the answer which most closely matches.

* 1MB
* **6MB**
* 25MB
* 87MB

In [87]:
!wget https://d37ci6vzurychx.cloudfront.net/trip-data/fhv_tripdata_2019-10.parquet

--2024-02-24 12:45:35--  https://d37ci6vzurychx.cloudfront.net/trip-data/fhv_tripdata_2019-10.parquet
Resolving d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)... 18.155.128.187, 18.155.128.46, 18.155.128.6, ...
Connecting to d37ci6vzurychx.cloudfront.net (d37ci6vzurychx.cloudfront.net)|18.155.128.187|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18344035 (17M) [application/x-www-form-urlencoded]
Saving to: ‘fhv_tripdata_2019-10.parquet’

fhv_tripdata_2019-1 100%[===================>]  17.49M  --.-KB/s    in 0.1s    

2024-02-24 12:45:35 (162 MB/s) - ‘fhv_tripdata_2019-10.parquet’ saved [18344035/18344035]



In [88]:
df_pandas = pd.read_parquet('fhv_tripdata_2019-10.parquet')
df_pandas.head()

dispatching_base_num     pickup_datetime    dropOff_datetime  PUlocationID  \
0               B00009 2019-10-01 00:23:00 2019-10-01 00:35:00         264.0   
1               B00013 2019-10-01 00:11:29 2019-10-01 00:13:22         264.0   
2               B00014 2019-10-01 00:11:43 2019-10-01 00:37:20         264.0   
3               B00014 2019-10-01 00:56:29 2019-10-01 00:57:47         264.0   
4               B00014 2019-10-01 00:23:09 2019-10-01 00:28:27         264.0   

   DOlocationID SR_Flag Affiliated_base_number  
0         264.0    None                 B00009  
1         264.0    None                 B00013  
2         264.0    None                 B00014  
3         264.0    None                 B00014  
4         264.0    None                 B00014

In [89]:
df_pandas.columns

Index(['dispatching_base_num', 'pickup_datetime', 'dropOff_datetime',
       'PUlocationID', 'DOlocationID', 'SR_Flag', 'Affiliated_base_number'],
      dtype='object')

In [90]:
df_pandas.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1897856 entries, 0 to 1897855
Data columns (total 7 columns):
 #   Column                  Dtype         
---  ------                  -----         
 0   dispatching_base_num    object        
 1   pickup_datetime         datetime64[ns]
 2   dropOff_datetime        datetime64[ns]
 3   PUlocationID            float64       
 4   DOlocationID            float64       
 5   SR_Flag                 object        
 6   Affiliated_base_number  object        
dtypes: datetime64[ns](2), float64(2), object(3)
memory usage: 101.4+ MB


In [92]:
df = spark.read \
    .parquet('fhv_tripdata_2019-10.parquet')

In [93]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B00009|2019-10-01 00:23:00|2019-10-01 00:35:00|       264.0|       264.0|   null|                B00009|
|              B00013|2019-10-01 00:11:29|2019-10-01 00:13:22|       264.0|       264.0|   null|                B00013|
|              B00014|2019-10-01 00:11:43|2019-10-01 00:37:20|       264.0|       264.0|   null|                B00014|
|              B00014|2019-10-01 00:56:29|2019-10-01 00:57:47|       264.0|       264.0|   null|                B00014|
|              B00014|2019-10-01 00:23:09|2019-10-01 00:28:27|       264.0|       264.0|   null|                B00014|
|     B00021         |2019-10-01 00:00:4

In [94]:
df.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropOff_datetime: timestamp (nullable = true)
 |-- PUlocationID: double (nullable = true)
 |-- DOlocationID: double (nullable = true)
 |-- SR_Flag: integer (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [130]:
#no lo usé
schema = types.StructType([
    types.StructField('dispatching_base_num', types.StringType(), True),
    types.StructField('pickup_datetime', types.TimestampType(), True),
    types.StructField('dropOff_datetime', types.TimestampType(), True),
    types.StructField('PUlocationID', types.DoubleType(), True),
    types.StructField('DOlocationID', types.DoubleType(), True),
    types.StructField('SR_Flag', types.IntegerType(), True),
    types.StructField('Affiliated_base_number', types.StringType(), True)
])

In [131]:
df = spark.read \
    .schema(schema) \
    .parquet('fhv_tripdata_2019-10.parquet')

In [132]:
df.printSchema()

root
 |-- dispatching_base_num: string (nullable = true)
 |-- pickup_datetime: timestamp (nullable = true)
 |-- dropOff_datetime: timestamp (nullable = true)
 |-- PUlocationID: double (nullable = true)
 |-- DOlocationID: double (nullable = true)
 |-- SR_Flag: integer (nullable = true)
 |-- Affiliated_base_number: string (nullable = true)



In [133]:
df.show()

+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|dispatching_base_num|    pickup_datetime|   dropOff_datetime|PUlocationID|DOlocationID|SR_Flag|Affiliated_base_number|
+--------------------+-------------------+-------------------+------------+------------+-------+----------------------+
|              B00009|2019-10-01 00:23:00|2019-10-01 00:35:00|       264.0|       264.0|   null|                B00009|
|              B00013|2019-10-01 00:11:29|2019-10-01 00:13:22|       264.0|       264.0|   null|                B00013|
|              B00014|2019-10-01 00:11:43|2019-10-01 00:37:20|       264.0|       264.0|   null|                B00014|
|              B00014|2019-10-01 00:56:29|2019-10-01 00:57:47|       264.0|       264.0|   null|                B00014|
|              B00014|2019-10-01 00:23:09|2019-10-01 00:28:27|       264.0|       264.0|   null|                B00014|
|     B00021         |2019-10-01 00:00:4

In [134]:
df = df.repartition(6)

In [135]:
df.write.parquet("fhv/2019/10/")

## Question 3:
Count records

How many taxi trips were there on the 15th of October?

Consider only trips that started on the 15th of October.

* 108,164
* 12,856
* 452,470
* **62,610**

In [136]:
df.createOrReplaceTempView('trips_data')

In [137]:
spark.sql("""
SELECT
    count(1)
FROM
    trips_data
""").show()

+--------+
|count(1)|
+--------+
| 1897856|
+--------+



In [138]:
spark.sql("""
SELECT
    count(1)
FROM
    trips_data
WHERE
    ((MONTH(pickup_datetime) = 10) AND (DAY(pickup_datetime) = 15) AND (YEAR(pickup_datetime) = 2019))
""").show()

+--------+
|count(1)|
+--------+
|   62629|
+--------+



In [139]:
df \
    .withColumn('pickup_date', F.to_date(df.pickup_datetime)) \
    .filter("pickup_date = '2019-10-15'") \
    .count()

62629

In [140]:
spark.sql("""
SELECT
    count(1)
FROM
    trips_data
WHERE
    to_date(pickup_datetime) = '2019-10-15'
""").show()

+--------+
|count(1)|
+--------+
|   62629|
+--------+



## Question 4:
Longest trip for each day

What is the length of the longest trip in the dataset in hours?

* 631,152.50 Hours
* 243.44 Hours
* 7.68 Hours
* **3.32 Hours**

In [159]:
df_time_duration = df\
                    .withColumn("pickup_date", F.to_date('pickup_datetime'))\
                    .withColumn("travel_duration", (F.col('dropOff_datetime')- F.col('pickup_datetime'))/60)

In [160]:
df_time_duration_2 = df\
                    .withColumn("pickup_date", F.to_date('pickup_datetime'))\
                    .withColumn("travel_duration", (F.col('dropOff_datetime').cast("long") - F.col('pickup_datetime').cast("long"))/60)

In [161]:
df_time_duration_2.select("travel_duration").show()

+------------------+
|   travel_duration|
+------------------+
| 47.56666666666667|
| 7.766666666666667|
|              39.0|
|11.466666666666667|
|              38.0|
|15.683333333333334|
|31.183333333333334|
|              25.0|
|17.316666666666666|
| 38.53333333333333|
|36.733333333333334|
|21.716666666666665|
|               3.1|
| 6.333333333333333|
| 4.466666666666667|
|16.616666666666667|
|21.066666666666666|
|17.983333333333334|
|              13.5|
|              3.25|
+------------------+
only showing top 20 rows



In [162]:
df_time_duration_2.select('pickup_date','travel_duration')\
                .groupBy('pickup_date')\
                .max('travel_duration')\
                .orderBy('max(travel_duration)', ascending = False)\
                .limit(5)\
                .show()

+-----------+--------------------+
|pickup_date|max(travel_duration)|
+-----------+--------------------+
| 2019-10-28|          3.786915E7|
| 2019-10-11|          3.786915E7|
| 2019-10-31|          5260346.45|
| 2019-10-01|   4207681.683333334|
| 2019-10-17|            527640.0|
+-----------+--------------------+



In [156]:
df_time_duration_2.createOrReplaceTempView('trips_data_duration')

In [189]:
spark.sql("""
SELECT 
    pickup_datetime,
    max(DATEDIFF(dropOff_datetime, pickup_datetime))
FROM
    trips_data
GROUP BY
    1
ORDER BY 
    2 DESC
LIMIT 5;
""").show()

+-------------------+------------------------------------------------+
|    pickup_datetime|max(datediff(dropOff_datetime, pickup_datetime))|
+-------------------+------------------------------------------------+
|2019-10-28 09:00:00|                                           26298|
|2019-10-11 18:00:00|                                           26298|
|2019-10-31 23:46:33|                                            3654|
|2019-10-01 21:43:42|                                            2922|
|2019-10-17 14:00:00|                                             367|
+-------------------+------------------------------------------------+



## Question 5:
User Interface

Spark’s User Interface which shows the application's dashboard runs on which local port?

* 80
* 443
* **4040**
* 8080

## Question 6:
Least frequent pickup location zone

Load the zone lookup data into a temp view in Spark
Zone Data

Using the zone lookup data and the FHV October 2019 data, what is the name of the LEAST frequent pickup location Zone?

* East Chelsea
* **Jamaica Bay**
* Union Sq
* Crown Heights North

In [70]:
!wget https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv

--2024-02-24 12:40:59--  https://github.com/DataTalksClub/nyc-tlc-data/releases/download/misc/taxi_zone_lookup.csv
Resolving github.com (github.com)... 140.82.121.3
Connecting to github.com (github.com)|140.82.121.3|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6ea97ed0e6a?X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAVCODYLSA53PQK4ZA%2F20240224%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20240224T124059Z&X-Amz-Expires=300&X-Amz-Signature=c891596a95e1f8da2aa9e078b0bb988c0dd5c2a46041593be276ad40d04cb385&X-Amz-SignedHeaders=host&actor_id=0&key_id=0&repo_id=513814948&response-content-disposition=attachment%3B%20filename%3Dtaxi_zone_lookup.csv&response-content-type=application%2Foctet-stream [following]
--2024-02-24 12:40:59--  https://objects.githubusercontent.com/github-production-release-asset-2e65be/513814948/5a2cc2f5-b4cd-4584-9c62-a6e

In [176]:
df_pandas_zone = pd.read_csv('taxi_zone_lookup.csv')

In [177]:
df_pandas_zone.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 265 entries, 0 to 264
Data columns (total 4 columns):
 #   Column        Non-Null Count  Dtype 
---  ------        --------------  ----- 
 0   LocationID    265 non-null    int64 
 1   Borough       265 non-null    object
 2   Zone          264 non-null    object
 3   service_zone  263 non-null    object
dtypes: int64(1), object(3)
memory usage: 8.4+ KB


In [178]:
df_zones = spark.read \
           .option("header", "true") \
           .csv('taxi_zone_lookup.csv')           

In [179]:
df_zones.printSchema()

root
 |-- LocationID: string (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [180]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [181]:
schema_zones = types.StructType([
    types.StructField('LocationID', types.LongType(), True),
    types.StructField('Borough', types.StringType(), True),
    types.StructField('Zone', types.StringType(), True),
    types.StructField('service_zone', types.StringType(), True),
])


In [182]:
df_zones = spark.read \
           .option("header", "true") \
           .schema(schema_zones) \
           .csv('taxi+_zone_lookup.csv')

In [183]:
df_zones.printSchema()

root
 |-- LocationID: long (nullable = true)
 |-- Borough: string (nullable = true)
 |-- Zone: string (nullable = true)
 |-- service_zone: string (nullable = true)



In [184]:
df_zones.show()

+----------+-------------+--------------------+------------+
|LocationID|      Borough|                Zone|service_zone|
+----------+-------------+--------------------+------------+
|         1|          EWR|      Newark Airport|         EWR|
|         2|       Queens|         Jamaica Bay|   Boro Zone|
|         3|        Bronx|Allerton/Pelham G...|   Boro Zone|
|         4|    Manhattan|       Alphabet City| Yellow Zone|
|         5|Staten Island|       Arden Heights|   Boro Zone|
|         6|Staten Island|Arrochar/Fort Wad...|   Boro Zone|
|         7|       Queens|             Astoria|   Boro Zone|
|         8|       Queens|        Astoria Park|   Boro Zone|
|         9|       Queens|          Auburndale|   Boro Zone|
|        10|       Queens|        Baisley Park|   Boro Zone|
|        11|     Brooklyn|          Bath Beach|   Boro Zone|
|        12|    Manhattan|        Battery Park| Yellow Zone|
|        13|    Manhattan|   Battery Park City| Yellow Zone|
|        14|     Brookly

In [185]:
df_zones.createOrReplaceTempView('trips_data_zones')

In [186]:
spark.sql("""
WITH trips as (
    SELECT *
    FROM trips_data
), 
trips_zones as (
    SELECT *
    FROM trips_data_zones
)
SELECT 
    trips.PULocationID,
    trips_zones.LocationID,
    trips_zones.Zone
FROM trips
INNER JOIN trips_zones
    on trips.PULocationID = trips_zones.LocationID
LIMIT 5;
""").show()

+------------+----------+-------------+
|PULocationID|LocationID|         Zone|
+------------+----------+-------------+
|        98.0|        98|Fresh Meadows|
|        98.0|        98|Fresh Meadows|
|        98.0|        98|Fresh Meadows|
|        98.0|        98|Fresh Meadows|
|        98.0|        98|Fresh Meadows|
+------------+----------+-------------+



In [187]:
spark.sql("""
SELECT 
    CONCAT(trips_zones_pup.Zone, "-", trips_zones_pup.Borough),
    COUNT(1)
FROM 
    trips_data trips LEFT JOIN trips_data_zones trips_zones_pup ON trips.PULocationID = trips_zones_pup.LocationID
GROUP BY
    1
ORDER BY
    2
LIMIT 5;
""").show()

+------------------------+--------+
|concat(Zone, -, Borough)|count(1)|
+------------------------+--------+
|      Jamaica Bay-Queens|       1|
|    Governor's Island...|       2|
|    Green-Wood Cemete...|       5|
|    Broad Channel-Queens|       8|
|    Highbridge Park-M...|      14|
+------------------------+--------+



In [188]:
spark.sql("""
SELECT
    PULocationID, 
    COUNT(1)
FROM
    trips_data
GROUP BY
    1
ORDER BY 
    2
LIMIT 5;
""").show()

+------------+--------+
|PULocationID|count(1)|
+------------+--------+
|         2.0|       1|
|       105.0|       2|
|       111.0|       5|
|        30.0|       8|
|       120.0|      14|
+------------+--------+

